src/bin/pg_autoctl/monitor.c

```diff
@@ -2994,16 +2994,54 @@ monitor_stop_maintenance(Monitor *monitor, int nodeId)
  * of notifications received from the previous calls loop.
  */
 bool
-monitor_get_notifications(Monitor *monitor)
+monitor_get_notifications(Monitor *monitor, int timeoutMs)
 {
 	PGconn *connection = monitor->pgsql.connection;
 	PGnotify *notify;
+
+	char *channels[] = { "state", "log", NULL };
+
+	int ret;
 	int sock;
+
+	sigset_t mask;
+	sigset_t orig_mask;
+
+	/* we have milliseconds, we want seconds and nanoseconds separately */
+	int seconds = timeoutMs / 1000;
+	int nanosecs = 1000 * 1000 * (timeoutMs % 1000);
+	struct timespec timeout = { .tv_sec = seconds, .tv_nsec = nanosecs };
+
 	fd_set input_mask;
 
-	if (connection == NULL)
+	/* block signals now: process them as if received during the pselect call */
+	if (!block_signals(&mask, &orig_mask))
+	{
+		return false;
+	}
+
+	if (asked_to_stop || asked_to_stop_fast || asked_to_reload || asked_to_quit)
+	{
+		/* restore signal masks (un block them) now */
+		(void) unblock_signals(&orig_mask);
+		return false;
```

Comment:
I think this should be:
```suggestion
		return true;
```

Reply:
When we have received a termination signal, it's could be that we want to say that the function did run successfully, but in the case of a reload we are going to disconnect, do things, and then reconnect. So we're going to miss notifications. I think it's fair to return false when we have signals to process in the main loop.

``` c
		if (!monitor_get_notifications(monitor,

		                               /* we want the time in milliseconds */
									   PG_AUTOCTL_MONITOR_SLEEP_TIME * 1000))
		{
			log_warn("Re-establishing connection. We might miss notifications.");
			pgsql_finish(&(monitor->pgsql));

			continue;
		}
```